### PR TITLE
Generate `Reservation` CRD

### PR DESCRIPTION
**Pull request by brycahta**
Issue #, if available: https://github.com/aws-controllers-k8s/community/issues/1196

Description of changes:
* Allows the creation and termination of multiple instances by defining a `Reservation` resource backed by `RunInstances` , `DescribeInstances`, and `TerminateInstances` apis

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by brycahta** [comment]
`kubectl describe reservations` sample output: https://gist.github.com/brycahta/66b1bea8c6f8810996344cc0bc0f0fee

This was a reservation for 2 instances and note the _Status_ section contains all instance details so kinda beefy.

File: templates/hooks/reservation/sdk_delete_post_build_request.go.tpl [hunk 1]
```diff
@@ -0,0 +1,3 @@
+    if err = addInstanceIDsToTerminateRequest(r, input); err != nil {
```
Comment (vijtrip2) [inline templates/hooks/reservation/sdk_delete_post_build_request.go.tpl:1]: Is it not possible to have zero instances in the reservation? Current code will error out of there are zero instances.

**Review by vijtrip2 — APPROVED** [review]

File: templates/hooks/reservation/sdk_delete_post_build_request.go.tpl (hunk 1 above)
Comment (brycahta) [reply templates/hooks/reservation/sdk_delete_post_build_request.go.tpl:1]: > Is it not possible to have zero instances in the reservation?

Correct. From the [RunInstancesInput docs](https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#RunInstancesInput) for both `MinCount` and `MaxCount` fields: 

```
Constraints: Between 1 and the maximum number you're allowed for the specified
    // instance type.
```


I'm planning to have the next PR add exceptions/terminal_codes to communicate this properly

File: apis/v1alpha1/reservation.go [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+	// matching attributes (instance type, platform, Availability Zone).
+	CapacityReservationSpecification *CapacityReservationSpecification `json:"capacityReservationSpecification,omitempty"`
+	// Unique, case-sensitive identifier you provide to ensure the idempotency of
+	// the request. If you do not specify a client token, a randomly generated token
+	// is used for the request to ensure idempotency.
+	//
+	// For more information, see Ensuring Idempotency (https://docs.aws.amazon.com/AWSEC2/latest/APIReference/Run_Instance_Idempotency.html).
+	//
+	// Constraints: Maximum 64 ASCII characters
+	ClientToken *string `json:"clientToken,omitempty"`
```
Comment (RedbackThomson) [inline apis/v1alpha1/reservation.go:48]: Please add to the ignore list

File: apis/v1alpha1/reservation.go [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+	PrivateIPAddress *string `json:"privateIPAddress,omitempty"`
+	// The ID of the RAM disk to select. Some kernels require additional drivers
+	// at launch. Check the kernel requirements for information about whether you
+	// need to specify a RAM disk. To find kernel requirements, go to the Amazon
+	// Web Services Resource Center and search for the kernel ID.
+	//
+	// We recommend that you use PV-GRUB instead of kernels and RAM disks. For more
+	// information, see PV-GRUB (https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/UserProvidedkernels.html)
+	// in the Amazon EC2 User Guide.
+	RamdiskID *string `json:"ramdiskID,omitempty"`
```
Comment (RedbackThomson) [inline apis/v1alpha1/reservation.go:209]: `RAM` should be added as a initialism

File: apis/v1alpha1/types.go [hunk 4]
```diff
@@ -1282,30 +1340,40 @@ type IPv6Range struct {
...
+	Name            *string        `json:"name,omitempty"`
+	OwnerID         *string        `json:"ownerID,omitempty"`
+	Platform        *string        `json:"platform,omitempty"`
+	PlatformDetails *string        `json:"platformDetails,omitempty"`
+	ProductCodes    []*ProductCode `json:"productCodes,omitempty"`
+	Public          *bool          `json:"public,omitempty"`
+	RamdiskID       *string        `json:"ramdiskID,omitempty"`
+	RootDeviceName  *string        `json:"rootDeviceName,omitempty"`
+	RootDeviceType  *string        `json:"rootDeviceType,omitempty"`
+	SriovNetSupport *string        `json:"sriovNetSupport,omitempty"`
```
Comment (RedbackThomson) [inline apis/v1alpha1/types.go:1363]: `SRIOV` should be added as an initialism

File: apis/v1alpha1/types.go [hunk 5]
```diff
@@ -1376,38 +1453,84 @@ type ImportVolumeTaskDetails struct {
...
+	CapacityReservationID *string                       `json:"capacityReservationID,omitempty"`
+	// Describes the instance's Capacity Reservation targeting preferences. The
+	// action returns the capacityReservationPreference response element if the
+	// instance is configured to run in On-Demand capacity, or if it is configured
+	// in run in any open Capacity Reservation that has matching attributes (instance
+	// type, platform, Availability Zone). The action returns the capacityReservationTarget
+	// response element if the instance explicily targets a specific Capacity Reservation
+	// or Capacity Reservation group.
+	CapacityReservationSpecification *CapacityReservationSpecificationResponse `json:"capacityReservationSpecification,omitempty"`
+	ClientToken                      *string                                   `json:"clientToken,omitempty"`
```
Comment (RedbackThomson) [inline apis/v1alpha1/types.go:1469]: Please add this to ignored

**Review by RedbackThomson — CHANGES_REQUESTED** [review]
This is really great. I love that you went with `Reservation` as the CRD name - although SEO might be a bit tough for documentation, so hopefully we can figure that out. I have some initialisms for you, and some field paths to ignore, but nothing crazy.

File: apis/v1alpha1/reservation.go [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+	PrivateIPAddress *string `json:"privateIPAddress,omitempty"`
+	// The ID of the RAM disk to select. Some kernels require additional drivers
+	// at launch. Check the kernel requirements for information about whether you
+	// need to specify a RAM disk. To find kernel requirements, go to the Amazon
+	// Web Services Resource Center and search for the kernel ID.
+	//
+	// We recommend that you use PV-GRUB instead of kernels and RAM disks. For more
+	// information, see PV-GRUB (https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/UserProvidedkernels.html)
+	// in the Amazon EC2 User Guide.
+	RAMdiskID *string `json:"ramdiskID,omitempty"`
```
Comment (RedbackThomson) [inline apis/v1alpha1/reservation.go:201]: What happened here? Why didn't it uppercase `Disk`?

File: apis/v1alpha1/types.go (hunk 5 above)
Comment (RedbackThomson) [reply apis/v1alpha1/types.go:1469]: This is coming from the Describe call, I believe. Please check

File: apis/v1alpha1/reservation.go (hunk 6 above)
Comment (brycahta) [reply apis/v1alpha1/reservation.go:201]: * RamdiskID --> RAMdiskID
* RamDiskID --> RAMDiskID

Comment (brycahta) [reply apis/v1alpha1/reservation.go:201]: I overlooked the first format.. didn't see it in my grep. Will update rules & regen.

File: apis/v1alpha1/reservation.go [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+)
+
+// ReservationSpec defines the desired state of Reservation.
+//
+// Describes a launch request for one or more instances, and includes owner,
+// requester, and security group information that applies to all instances in
+// the launch request.
+type ReservationSpec struct {
+	// Reserved.
+	AdditionalInfo *string `json:"additionalInfo,omitempty"`
```
Comment (RedbackThomson) [inline apis/v1alpha1/reservation.go:30]: Ummm should this be ignored? It sounds like we can't use it anyway?

**Comment by brycahta** [comment]
Discussed offline; renaming the resource to `Instance` and hard-coding min/max count to 1, so a single instance is created and can be managed/modified by the controller (i.e. able add volume to instance). If we proceed with `Reservation`, then there is no way to update the reservation/collection of instances in this way.

**Comment by brycahta** [comment]
Stale; re-genning as `Instance` once https://github.com/aws-controllers-k8s/community/issues/1285 is resolved

**Pull request closed without merging**